New Markdown Editor implementation with live preview

paulocheque opened the pull request:
Using Ace Editor + Marked + HighlightJS + DropZone.

New features:
- Live preview
- Markdown highlight
- Code highlight
- Scroll sync
- Autosave the tutorial in the browser local storage
- Header auto update
- Link to upload files
- DropZone text
- New degisn
- Line numbers
- Live Markdown tutorial by example
- New external references for Markdown tutorials

I am creating this PR now, but I can improve/update it after feedbacks.

The autosave feature can be improved. We can clear the local cache after the tutorial has been successfully submitted. But we will need to refactoring not just the JS function, but the server endpoint too.

paulocheque committed:
Added dev folders into .gitignore

paulocheque committed:
Using Epic Editor. Features included:
- Live and side by side preview
- Design close to the mockup
- Using local auto save per article or for draft
- Auto update preview header

paulocheque committed:
Auto focus on article title

paulocheque committed:
Using DropZone to accept drag and drop into the Epic Editor.

paulocheque committed:
Fixed typo in the editor page. Using hack.guides logo in the editor.

paulocheque committed:
Removing useless html padding in the editor page: stop including internal iframe css files in the editor to avoid conflicts.

paulocheque committed:
Browser's local storage renamed to `articles`.

paulocheque committed:
Fixed css/js libs import after rebase

paulocheque committed:
Autosave changed to 200ms

paulocheque committed:
Using Ace Editor + Marked + HighlightJS + DropZone.

paulocheque committed:
Sending the article content when saving the article with the new Ace Editor

paulocheque committed:
Update preview on the page load. Also, new placeholder text for drafts.

paulocheque committed:
Editor's placeholder text updated with link examples

paulocheque committed:
Better sync scroll strategy, using percentages of the editor/preview's height, since they might have difference heights. Also, sync scroll even if the user scroll the preview area.

paulocheque committed:
Wrap mode enabled in the editor to avoid horizontal scrolling. Also, setting a limit of 99999 lines.

paulocheque committed:
Fixed line breaks issue on submit the article. Post encoding with input remove the line breaks.

paulocheque committed:
Fixed issue on the first line of the preview

paulocheque committed:
Stop selecting all editor text on load. Move the cursor to the first line.

paulocheque committed:
Removed Ace Editor warning

paulocheque committed:
Fixed loading of the article content during the article's edition. Manually escaping special chars.

paulocheque committed:
Added autosave in the Ace Editor. Increase the timeout number to improve the performance for large files.

paulocheque committed:
More Markdown examples in the placeholder text: Tables (with design updated), horizontal rules, strikethrough and nested lists.

paulocheque committed:
Linking better markdown tutorials

paulocheque committed:
Using GitHub line breaks. gfm.

paulocheque committed:
Better markdown tutorial placeholder organisation

durden commented:
Also, @paulocheque setup a running example of this pull request if anyone else wants to try it out.  You can access that on [this test site](http://guides-dev.herokuapp.com).

durden commented:
Things look awesome so far.  I'm going to loop in @prtkgpt and @edroman to get their thoughts.  Below are some small issues I've noticed in recent testing. Don't let the length of this post fool you; this is great work and a lot of this is fairly minor.
1. Local storage isn't cleared, which leads to incorrectly filling in the editor after I submitted something.
   - For example, I successfully saved a guide yesterday then came back today via the 'write guide' link.  I intended to start a new guide, but the guide I submitted yesterday was filled into the editor.
   - Maybe we could provide an explicit load button on the editor to avoid this?
   - The idea being we never automatically fill in the editor with local storage contents unless the users asks for it.  This might be easier than re-architecting things to clear up local storage.  I'm not sure about how difficult that is.
2. The preview pane has a grey background, which is not how the article renders on the _real_ article pages.
   - I see the point of having the editing and preview panes different colors, but having a grey background on the preview makes syntax highlighting, etc. look different.  The goal of the preview is to show what the final guide will look like as close as possible.
   - Let's try having a white background for the both the editing and preview pane.
3. Scroll sync doesn't take into account the preview header and relative spacing of headers.
   - The scroll sync is a very powerful feature, but it's a bit off right now.  The sync seems to be based on the absolute position in each editor pane, not the relative position.  For example, rendered header text has more vertical spacing than the raw markdown text.  So, scrolling in the preview pane needs to be offset for this somehow.
   - The idea is that what the author sees in raw markdown editor pane matches the rendered text in the preview pane.  Obviously this probably won't always be possible because the panes are both the same height and the vertical spacing in the panes are different by necessity.  However, maybe we can improve this a little bit.
   - **Are there any other editors out there we could try this on, maybe this is a solved problem or a commonly accepted 'solution'?**
   - See the attached image for a better example.  Notice how the top of the preview pane is 'Python tools', but the editor pane is not showing that text at all.
   - Finally, the scroll sync needs to take into account the larger 'header' portion of the preview.  Notice how the preview pane has the title and author name, but the editor pane doesn't.  This makes sense, but the scroll sync needs to be aware of that and scroll accordingly.
   - Does this make sense?  Is this difficult to fix?
     <img width="1194" alt="editor_scrolling" src="https://cloud.githubusercontent.com/assets/58063/14148460/d6be48d6-f69f-11e5-9876-a5bedf6921d8.png">
4. Need a full-screen button.
   - I'm not sure if the Ace editor has full-screen built-in or not.  It might not be too important to add at this point if the capability is not built-in.
5. Should there be a way to stretch the height of the editor?
   - This could be similar to Github's UI where you can drag the bottom right corner to make the editor window larger.
   - This may be unnecessary if we have full-screen capability.
6. Should we add a button to easily clear the text?
   - The starter text in the editor is great for someone who is new to markdown or seeing the editor for the first time.  However, each time an author writes a new guide they have to select-all and delete this placeholder text. Maybe we could provide a button to do this easily?
   - This could be slightly dangerous in case a user accidentally hits this.  So, we probably want to be careful about how we save to local storage when this button is hit.  This way we could easily retrieve the contents of local storage for an 'undo' button.

paulocheque commented:
Hi, ok. I have already describe some of these comments to you but I will put them here to unify the conversation.

1) We would need to store the local storage until the user successfully save the tutorial in our servers. After that, the local cache is useless. So, I don't think it is necessary to add new buttons because it will increase the complexity of the system. If we can track the form submission (it need some refactoring in the backend and in the frontend), we can solve this.

2) Cool!

3) Independent of the header, the height of the preview will always be different of the editor. For that reason I stopped to use the absolute position. For example: when we are in the end of one side, we will not be in the end of the another side.

I am using the same strategy of Dillinger's editor. Using percentages. 

I guess this problem is hard to be solved automatically. For that reason the scroll sync has to be enabled/disabled by the user.

4) Ace editor does not provide this feature out of the box. Maybe this can be splitted in another feature/PR. I guess this feature will not be complicated to implement with the current implementation.

5) I guess this would not be simple to implement and this could generate lots of design issues. I think it would be better to use the fullscreen feature instead.

6) I thought this too. The use can always `control+z` to back any change. But yes, the local storage would be cleaned too. But if the user undo, the local storage will be back again.

Another option would be to have another kind of button instead of the "clear button": The "Load Tutorial into editor" button.

durden commented:
Good point.  I assume the scroll sync would be complicated.  I just tested it in dillinger.io and you're right.  They have the same issue.  However, ours is worse because our preview shows the header and author information, which is not in the editor pane at all.  Maybe we could 'pad' the scroll sync to take this into account so the syncing is a little closer?

Another approach is to just not show the header and author information in the preview.  Anyone disagree with that?

durden commented:
4) Lets stick with another issue/pull request for full-screen editing.
5) Good point. I prefer the full-screen approach too.
6) Yea, I like your suggestion.  Maybe we should make the default be empty text and have a 'load tutorial' button to populate the sample text.
    - Maybe we could also have a 'clear tutorial' button that just detects if the tutorial text is shown and deletes it?

durden commented:
1) On second thought I think we need this fixed before we push this editor to the live site.  The reason is authors can get very confused if they write/edit a guide, then later come back to start a new one and get the old text.

What are your thoughts on refactoring the save endpoint?

paulocheque commented:
1) It makes perfect sense to me. I will focus on that.

3) I can improve the padding to minimize the problem. But is this header really necessary? Maybe we can try the solution of the following picture:

![screen shot 2016-03-30 at 2 10 51 pm](https://cloud.githubusercontent.com/assets/94350/14150895/abaef9ba-f681-11e5-96ea-f5c5a03a6df1.png)

6) I understand your idea and it is cool! I will create something like that: some "playground mode" or "tutorial mode".

durden commented:
3) Great idea lets go with your suggestion. The header really isn't that
necessary so we may end up removing it anyway but your suggestion is a nice
step in the right direction.

On Wednesday, March 30, 2016, Paulo Cheque notifications@github.com wrote:

> 1) It makes perfect sense to me. I will focus on that.
> 
> 3) I can improve the padding to minimize the problem. But is this header
> really necessary? Maybe we can try the solution of the following picture:
> 
> [image: screen shot 2016-03-30 at 2 10 51 pm]
> https://cloud.githubusercontent.com/assets/94350/14150895/abaef9ba-f681-11e5-96ea-f5c5a03a6df1.png
> 
> 6) I understand your idea and it is cool! I will create something like
> that: some "playground mode" or "tutorial mode".
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/pluralsight/guides-cms/pull/30#issuecomment-203535430

paulocheque committed:
No more preview header text. Using preview header area to include the article's title and stack.

paulocheque committed:
Live Markdown tutorial feature with autosave disabled. User can enable/disable the tutorial at any time.

paulocheque commented:
3) I made changes to try the new design. In the production version, there is an issue in the author link (it points to another page instead of the author's page). Check the dev env.

6) I created a new button "live markdown tutorial" and the user can enable/disable this button at any time without affecting his tutorial and the saved content.

paulocheque committed:
New editor/preview background colours: white

paulocheque commented:
2) updated

durden commented:
Found another small tweak we should make.  Don't autofocus on the title if the user is editing an existing guide.  You can determine this by seeing if the template variable `article` is None.  The reason for this is we currently have editing of titles disabled so there's no reason to autofocus on an element users cannot change.

durden commented:
The white preview background looks much better, thanks!
- The live tutorial text has few typos with the word 'bloquote' which should be spelled 'blockquote'.
- I don't see a link for the author showing up in the new 'header' section of the preview bar.  Lets just remove the author information from there if you have it showing up now.  This doesn't add any value so might as well simplify.
- Can you increase the size of the title font in the preview 'header' area?  Lets try something like `20px`.
- Could we center the title text in the preview 'header' area?  Maybe this wouldn't leave so much whitespace for short titles.  Any pros/cons to this?

durden commented:
I've hit a weird bug that seems reproducible in Safari at least.  Try the following:
- Go to the write page to create a new guide
- Focus on the editor window
- Select all to copy and then delete the starter text
- Go to title and type something in
- Now I cannot focus back on the editor by clicking it
- I must either tab over to the editor or click on the single empty line to get focus and start typing.
- It would be good if clicking anywhere in the editor gave it focus and you could start typing.

See screenshot for better explanation.
<img width="1220" alt="editor_focus" src="https://cloud.githubusercontent.com/assets/58063/14178270/2d9b7024-f75a-11e5-9a68-fd68fc1f51c9.png">

I thought things were broken for a while because I didn't realize I had to click my mouse on the single blank line instead of anywhere in the editor.

paulocheque committed:
autofocus only for new articles

paulocheque committed:
Fixed typo

paulocheque committed:
Editor must always fill the entire area. This facilitates focusing the editor too.

paulocheque committed:
Forms improvements

paulocheque committed:
Preview header title centered. Design updated.

paulocheque committed:
Always give preference for the local cache of the article. Fixed issue loading the bootstrap live tutorial for saved articles.

paulocheque commented:
All of that are fixed. It is pending just the submission handler, and I am currently working on it.

paulocheque committed:
Refactoring /save endpoint to become /api/save endpoint. It will return a json data with pertinent status code to be used by the frontend. Adapting the frontend to use the new endpoint and deal with the request result.

durden commented on pskb_website/static/js/editor_utils.js:
```diff
@@ -317,6 +317,7 @@ function save(sha, path, secondary_repo, action_url) {
     var stacks = document.createElement("input");
     stacks.name = "stacks";
     stacks.value = stacks_select.value;
+    stacks.required = "required";
```
What does this do?  I tested it and selecting a stack is not required.  We don't want to require a stack since we have a finite list, so just curious what this line does.

Ok I see what's going on guess we're still translating the '---' stack with a value of "" correctly on the server-side.  So `required` just means there has to be 'some' value with stack and it has to show up in the form?  Just trying to learn here so thanks for the info!

paulocheque replied:
I was just trying to improve the form, since I saw in the frontend we discard empty values for the stack.

But we can ignore this piece of code since I replaced this with the new version using Jquery.Ajax

durden commented:
Awesome all of those changes are great.  Thanks!

paulocheque committed:
Clear local cache of the article before successfully saving the article into our servers.

paulocheque commented:
I refactored the backend and the form submission:
- /save changed to /api/save
- backend: refactored to return a informative json data with informative http code
- frontend: refactored to use jquery.ajax to handle the form submission

Also, now I am cleaning the local cache after succesfully form submission.

paulocheque committed:
Disabling save button if the live markdown tutorial is opened. Also, disabling it during form submission.

paulocheque commented:
ps: I rebased my fork with the origin master to keep it all updated. So the commits are together in this thread now.

durden commented:
Thanks, the changes look pretty good.  I think your fork was still a bit out of date based on a [few tiny changes](https://github.com/pluralsight/guides-cms/commit/dc89a71eab1c46f6c754850a2526c24be5814554) I made on Friday.  Could you pull in those changes too?  I think it will merge cleanly then.  It looks pretty small, but I don't want to break any of your changes when merging.

My plan is to do a bit more testing Monday/Tuesday and then roll this into the main repo and deploy to the production site.  Does this sound OK @paulocheque?

Then, we'll use another Pull Request to cleanup the design of the editor page itself, maybe using a 2 page workflow, etc.  Those changes should be relatively minor though compared to this and probably only HTML/CSS and maybe a bit of javascript.  Let me know if anyone else has other ideas or suggestions.

durden commented on pskb_website/views.py:
```diff
@@ -628,12 +635,18 @@ def save():
+        data = {'error': 'Failed creating guide on github', 'redirect': redirect_to}
+        return Response(response=json.dumps(data), status=500, mimetype='application/json')
 
+    # TODO: move this to another endpoint
```
Good point.  I've added a new [issue](https://github.com/pluralsight/guides-cms/issues/33) to refactor this out of the main save endpoint.

durden commented on pskb_website/views.py:
```diff
@@ -546,20 +546,22 @@ def partner(article_path):
 
     if user.email is None:
         flash('Unable to read email address from Github API to properly attribute your commit to your account. Please make sure you have authorized the application to access your email.', category='warning')
+        # FIXME: stop using flash
```
What would you suggest we do here? Maybe save off this message and return it with response?

durden commented:
@paulocheque The conflict was really small and you made a change to return 401 which I also made on Friday, guess we were thinking the same thing!

I've merged the changes and have this running on our [staging site](http://pskb-stage.herokuapp.com/) for more testing after the merge.  @edroman @prtkgpt @paulocheque Can you guys all give it a quick test before I roll it into the production site?

I've tested it myself, and the changes seem OK.  We'll need to tweak the the editor page itself for a better UX, but I think we'll be getting some designer input on that later this week.

Also, I switched scroll sync to be the default as @edroman previously suggested.  Let me know if you guys have any questions or concerns.  I'll roll this to the production site first thing Tuesday morning barring any testing issues you might have.

paulocheque commented:
Cool that everything is fine so far. I will make tests in the staging env.

paulocheque commented:
It is everything fine!
